worker/caasoperator/caasoperator.go

```diff
@@ -255,17 +261,48 @@ func (op *caasOperator) loop() (err error) {
 		return errors.Trace(err)
 	}
 
-	aliveUnits := make(set.Strings)
+	// Keep a record of the alive units an a channel used to notify
+	// their uniter workers when the charm version has changed.
+	aliveUnits := make(map[string]chan struct{})
 
 	if err = restartWatcher(); err != nil {
 		err = errors.Annotate(err, "(re)starting watcher")
 		return errors.Trace(err)
 	}
 
+	// We should not do anything until there has been a change
+	// to the remote state. The watcher will trigger at least
+	// once initially.
+	select {
+	case <-op.catacomb.Dying():
+		return op.catacomb.ErrDying()
+	case <-watcher.RemoteStateChanged():
```

Comment:
You have a situation here where this agent may be down when the upgrade is requested. When the agent starts, you are consuming the first event without checking the state of the world.

Reply:
This is fine. What happens is the operator will always download and unpack the latest charm when it starts. The uniter worker have logic to check the state of the world (they also swallow this initial event) and will react corrrectly.